File: clojure/ch02/ex2.61.clj
(ns sicp.ch02 (:use clojure.test))

; Set representation with ordered entries (assume numbers) in the list.

(defn element-of-set? [x set]
  (cond (empty? set) false
        (= x (first set)) true
        (< x (first set)) false
        :else (element-of-set? x (rest set))))
          
(deftest test-element-of-set?
  (is (= false (element-of-set? 1 '())))
  (is (= false (element-of-set? 1 '(2 3 4))))
  (is (= true  (element-of-set? 1 '(1 2 3 4))))
  (is (= true  (element-of-set? 1 '(1)))))

(defn append [list1 list2]
  (loop [l1 (reverse list1) l2 list2]
    (if (empty? l1)
        l2
        (recur (rest l1) (cons (first l1) l2)))))

; This is also O(n), but on average takes 1/2 the time of adjoin-set in an
; unordered list implementation, just like element-of-set? above. This is true
; because the traversal of the existing set will stop when one of the two
; conditionals (= or <) is true.
(defn adjoin-set [x set]
  (if (empty? set)
      (list x)
      (let [s (first set)]
        (cond (= x s) set
              (< x s) (cons x set)
              :else (cons s (adjoin-set x (rest set)))))))

(deftest test-adjoin-set
  (is (= '(1) (adjoin-set 1 '())))
  (is (= '(1) (adjoin-set 1 '(1))))
  (is (= '(1 2) (adjoin-set 1 '(2))))
  (is (= '(1 2) (adjoin-set 1 '(1 2))))
  (is (= '(1 2 3 4) (adjoin-set 1 '(1 2 3 4))))
  (is (= '(1 2 3 4) (adjoin-set 2 '(1 2 3 4))))
  (is (= '(1 2 3 4) (adjoin-set 3 '(1 2 3 4))))
  (is (= '(1 2 3 4) (adjoin-set 4 '(1 2 3 4))))
  (is (= '(1 2 3 4) (adjoin-set 1 '(2 3 4))))
  (is (= '(1 2 3 4) (adjoin-set 2 '(1 3 4))))
  (is (= '(1 2 3 4) (adjoin-set 3 '(1 2 4))))
  (is (= '(1 2 3 4) (adjoin-set 4 '(1 2 3)))))

(defn intersection-set [set1 set2]
  (if (or (empty? set1) (empty? set2))
    '()
    (let [x1 (first set1) x2 (first set2)]
      (cond (= x1 x2)
              (cons x1 (intersection-set (rest set1) (rest set2)))
            (< x1 x2)
              (intersection-set (rest set1) set2)
            (< x2 x1)
              (intersection-set set1 (rest set2))))))

(deftest test-intersection-set
  (is (= '() (intersection-set '() '())))
  (is (= '() (intersection-set '() '(1 2 3))))
  (is (= '() (intersection-set '(1 2 3) '())))
  (is (= '(2) (intersection-set '(2) '(2 3 4 5))))
  (is (= '(2) (intersection-set '(2 3 4 5) '(2))))
  (is (= '(2 3) (intersection-set '(2 3 4 5) '(1 2 3))))
  (is (= '(2 3) (intersection-set '(1 2 3) '(2 3 4 5)))))

(run-tests)

File: scala/ch02/ex2.61.scala
// Set representation with ordered entries in the list.

def elementOfSet (i:Int, set: List[Int]): Boolean = set match {
  case Nil => false
  case head::tail => 
    if      (head == i) true
    else if (head >  i) false
    else elementOfSet (i, tail)
}

import org.scalatest._ 
import org.scalatest.matchers._

object elementOfSetSpec extends Spec with ShouldMatchers {
  describe ("elementOfSet") {
    it ("should return false if the item is not in the set") {
      elementOfSet (1, Nil) should equal (false)
      elementOfSet (1, List(2,3,4)) should equal (false)
    }
    it ("should return true if the item is in the set") {
      elementOfSet (1, List(1,2,3,4)) should equal (true)
      elementOfSet (1, List(1)) should equal (true)
    }
  }
}
elementOfSetSpec execute

// This is also O(n), but on average takes 1/2 the time of adjoin-set in an
// unordered list implementation, just like element-of-set? above. This is true
// because the traversal of the existing set will stop when one of the two
// conditionals (= or <) is true.
def adjoinSet (i:Int, set: List[Int]): List[Int] = {
  if (set == Nil)
    List (i)
  else {
    val s = set.head
    if      (i == s) set
    else if (i < s) i :: set
    else s :: adjoinSet(i, set.tail)
  }
}
  
object adjoinSetSpec extends Spec with ShouldMatchers {
  describe ("adjoinSet") {
    it ("should return a new set with new element as the head and the old set as the tail") {
      adjoinSet (1, Nil) should equal (List(1))
      adjoinSet (1, List(1)) should equal (List(1))
      adjoinSet (1, List(2)) should equal (List(1,2))
      adjoinSet (1, List(1,2)) should equal (List(1,2))
      adjoinSet (1, List(1,2,3,4)) should equal (List(1,2,3,4))
      adjoinSet (2, List(1,2,3,4)) should equal (List(1,2,3,4))
      adjoinSet (3, List(1,2,3,4)) should equal (List(1,2,3,4))
      adjoinSet (4, List(1,2,3,4)) should equal (List(1,2,3,4))
      adjoinSet (1, List(2,3,4)) should equal (List(1,2,3,4))
      adjoinSet (2, List(1,3,4)) should equal (List(1,2,3,4))
      adjoinSet (3, List(1,2,4)) should equal (List(1,2,3,4))
      adjoinSet (4, List(1,2,3)) should equal (List(1,2,3,4))
    }
  }
}
adjoinSetSpec execute

def intersectionSet (set1: List[Int], set2: List[Int]): List[Int] = {
  if      (set1 == Nil || set2 == Nil) Nil
  else if (elementOfSet (set1.head, set2))
    set1.head :: intersectionSet (set1.tail, set2)
  else 
    intersectionSet (set1.tail, set2)
}

object intersectionSetSpec extends Spec with ShouldMatchers {
  describe ("intersectionSet") {
    it ("should return the intersection of two sets") {
      intersectionSet (Nil, Nil) should equal (Nil)
      intersectionSet (Nil, List(1,2,3)) should equal (Nil)
      intersectionSet (List(1,2,3), Nil) should equal (Nil)
      intersectionSet (List(2), List(2,3,4,5)) should equal (List(2))
      intersectionSet (List(2,3,4,5), List(2)) should equal (List(2))
      intersectionSet (List(2,3,4,5), List(1,2,3)) should equal (List(2,3))
      intersectionSet (List(1,2,3), List(2,3,4,5)) should equal (List(2,3))
    }
  }
}
intersectionSetSpec execute
